### PR TITLE
Enable user case creation for Web Users

### PR DESCRIPTION
**Pull request by dominicochoa-dimagi**
## Product Description

These changes enable the creation of user cases for web users. When a web user is saved after signup, a user case is then created with the relevant info. 

~~As this PR stands currently, the changes result in user cases being saved after user creation. However, the user case~~
~~does not appear in the case list. This is true for both "web-users" and "commcare-users."~~

~~Before we merge this PR the above issue needs to be resolved.~~

This was just a local ES issue and was fixed with `./manage.py ptop_reindexer_v2 sql-case`

## Technical Summary

[Jira Ticket](https://dimagi-dev.atlassian.net/browse/USH-943?atlOrigin=eyJpIjoiNDhmNmU2NzE4OWQzNDE5Nzk3ZGNiMjBmMTJkNjc0MWUiLCJwIjoiaiJ9)

Previously, the `sync_usercases` code was dependent on the `CommCareUser` model object. There are several feature flag checks throughout the code that call `CommCareUser.project` to retrieve the relevant domain. 

Because `WebUser` objects do not have a `project` property containing the domain name, the first issue was retrieving the relevant domain(s). `sync_webuser_usercases_if_applicable` was created which uses the `WebUser.get_domains` property to iterate over all of the domains the web user has membership. 

```python
def sync_webuser_usercases_if_applicable(user_id, spawn_task):
    user = CouchUser.get_by_user_id(user_id)
    for domain in user.get_domains():
        domain_obj = Domain.get_by_name(domain)
        if domain_obj.call_center_config.enabled or domain_obj.usercase_enabled:
            if spawn_task:
                sync_webuser_usercases_task.delay(user, domain_obj)
            else:
                sync_webuser_usercases_task(user, domain_obj)
```

The next issue was addressing the need to change the downstream feature flag checks to use a separate domain object rather than calling `CommCareUser.project`. A bulk of the changes for this ticket revolve around changing some of the helper functions to now have a `domain_obj` argument which is then used for the needed checks.

Thus: 

```python
def _iter_sync_usercase_helpers(user):
    if user.project.usercase_enabled:
        yield _get_sync_usercase_helper(
            user,
            USERCASE_TYPE,
            user.get_id
        )
```

becomes:

```python
def _iter_sync_usercase_helpers(user, domain_obj):
    usercase_type = WEB_USER_USERCASE_TYPE if user.is_web_user() else USERCASE_TYPE 
    if domain_obj.usercase_enabled:
        yield _get_sync_usercase_helper(
            user,
            domain_obj,
            usercase_type,
            user.get_id
        )
```

## Feature Flag

usercases & USH_web_user_case_creation (new)

### Safety story

Works safely locally and wrote test coverage in an already existing unit test. Also updated some other unit tests that were mildly effected. Changes are limited to usercase syncing. 

### Automated test coverage

https://github.com/dimagi/commcare-hq/blob/49d71dae375980a4e967d554cdb7b1142f87d3fa/corehq/form_processor/tests/test_usercase_dbaccessor.py#L11-L35

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/users/models.py [hunk 1]
```diff
@@ -1453,8 +1453,10 @@ def save(self, fire_signals=True, **params):
 
         if fire_signals:
             from .signals import couch_user_post_save
+            from corehq.apps.callcenter.tasks import sync_webuser_usercases_if_applicable
             results = couch_user_post_save.send_robust(sender='couch_user', couch_user=self)
             log_signal_errors(results, "Error occurred while syncing user (%s)", {'username': self.username})
+            sync_webuser_usercases_if_applicable(self._id, spawn_task=True)
```
Comment (snopoke) [inline corehq/apps/users/models.py:1459]: since this is called directly you might as well pass in the user object instead of just the ID. That will save having to fetch it again.

File: corehq/apps/callcenter/tasks.py [hunk 2]
```diff
@@ -46,6 +46,17 @@ def calculate_indicators():
...
 
+def sync_webuser_usercases_if_applicable(user_id, spawn_task):
+    user = CouchUser.get_by_user_id(user_id)
+    for domain in user.get_domains():
+        domain_obj = Domain.get_by_name(domain)
+        if domain_obj.call_center_config.enabled or domain_obj.usercase_enabled:
+            if spawn_task:
+                sync_webuser_usercases_task.delay(user, domain_obj)
+            else:
+                sync_webuser_usercases_task(user, domain_obj)
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:57]: Might as well just call `sync_usercases(user, domain_obj)` directly.

File: corehq/apps/callcenter/tasks.py [hunk 3]
```diff
@@ -46,6 +46,17 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_webuser_usercases_if_applicable(user_id, spawn_task):
+    user = CouchUser.get_by_user_id(user_id)
+    for domain in user.get_domains():
+        domain_obj = Domain.get_by_name(domain)
+        if domain_obj.call_center_config.enabled or domain_obj.usercase_enabled:
+            if spawn_task:
+                sync_webuser_usercases_task.delay(user, domain_obj)
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:55]: Here you want to pass in the User ID and have the task re-fetch the user when it runs. This has 2 advantages:
1. the task get's a fresh copy of the user when it runs
2. we don't need to serialize the whole user object for the task

File: corehq/apps/callcenter/tasks.py [hunk 4]
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
...
     if (user.project.call_center_config.enabled or user.project.usercase_enabled):
         if spawn_task:
-            sync_usercases_task.delay(user._id)
+            sync_usercases_task.delay(user._id, user.project)
         else:
-            sync_usercases_task(user._id)
+            sync_usercases_task(user._id, user.project)
+
+
+@task(serializer='pickle', queue='background_queue')
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:77]: once you're just passing in the user ID you can remove the custom serialization here:
```suggestion
@task(queue='background_queue')
```

File: corehq/apps/callcenter/tasks.py [hunk 5]
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
         return
     if (user.project.call_center_config.enabled or user.project.usercase_enabled):
         if spawn_task:
-            sync_usercases_task.delay(user._id)
+            sync_usercases_task.delay(user._id, user.project)
         else:
-            sync_usercases_task(user._id)
+            sync_usercases_task(user._id, user.project)
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:74]: Just call the wrapped function directly
```suggestion
            sync_usercases(user, user.project)
```

File: corehq/apps/callcenter/tasks.py [hunk 6]
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
         return
     if (user.project.call_center_config.enabled or user.project.usercase_enabled):
         if spawn_task:
-            sync_usercases_task.delay(user._id)
+            sync_usercases_task.delay(user._id, user.project)
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:72]: Similar to above, pass in the domain name rather than the whole domain:
```suggestion
            sync_usercases_task.delay(user._id, user.domain)
```

File: corehq/apps/callcenter/tasks.py [hunk 7]
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
...
+            sync_usercases_task(user._id, user.project)
+
+
+@task(serializer='pickle', queue='background_queue')
+def sync_webuser_usercases_task(user, domain_obj):
+    sync_usercases(user, domain_obj)
+    return None
 
 
 @task(serializer='pickle', queue='background_queue')
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:83]: once domain name is being passed in instead of the whole domain:
```suggestion
@task(queue='background_queue')
```

**Review by snopoke — COMMENTED** [review]
Definitely on the right track here. One thing to mention is that I don't think we should support 'call-center' features for Web Users.

File: corehq/apps/callcenter/tasks.py [hunk 8]
```diff
@@ -46,6 +46,17 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_webuser_usercases_if_applicable(user_id, spawn_task):
```
Comment (orangejenny) [inline corehq/apps/callcenter/tasks.py:49]: Nitpick: make `web_user` two words

File: corehq/apps/callcenter/sync_usercase.py [hunk 9]
```diff
@@ -93,18 +93,17 @@ def _domain_has_new_fields(domain, field_names):
     return False
 
 
-def _get_sync_usercase_helper(commcare_user, case_type, owner_id, case=None):
-    domain = commcare_user.domain
-    fields = _get_user_case_fields(commcare_user, case_type, owner_id)
-    case = case or CaseAccessors(domain).get_case_by_domain_hq_user_id(commcare_user._id, case_type)
-    close = commcare_user.to_be_deleted() or not commcare_user.is_active
-    user_case_helper = _UserCaseHelper(domain, owner_id, commcare_user._id)
+def _get_sync_usercase_helper(user, domain_obj, case_type, owner_id, case=None):
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/callcenter/sync_usercase.py:96]: Agree with this, just wanted to show that it's fine to re-fetch domain objects all over the place, because we cache `Domain.get_by_name`: https://github.com/dimagi/commcare-hq/blob/fc8ca8faa838f5bbee50d164147e5b3078fbfe41/corehq/apps/domain/models.py#L607-L610

File: corehq/apps/callcenter/tasks.py [hunk 10]
```diff
@@ -58,12 +69,18 @@ def sync_usercases_if_applicable(user, spawn_task):
...
+            sync_usercases_task.delay(user._id, user.project)
         else:
-            sync_usercases_task(user._id)
+            sync_usercases_task(user._id, user.project)
+
+
+@task(serializer='pickle', queue='background_queue')
+def sync_webuser_usercases_task(user, domain_obj):
+    sync_usercases(user, domain_obj)
+    return None
```
Comment (orangejenny) [inline corehq/apps/callcenter/tasks.py:80]: nitpick: don't need the explicit `return None`

**Comment by dominicochoa-dimagi** [comment]
Made the suggested changes. User cases are still not populating in 'Case List.' Any suggestions?

File: corehq/apps/callcenter/tasks.py [hunk 11]
```diff
@@ -46,6 +46,16 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_web_user_usercases_if_applicable(user, spawn_task):
```
Comment (orangejenny) [inline corehq/apps/callcenter/tasks.py:49]: Should this function also check `settings.UNIT_TESTING` like `sync_usercases_if_applicable` does?

This and `sync_usercases_if_applicable` are right on the borderline where it might be better to collapse them into one function. `user.get_domains()` should also work for mobile users, it'll just return their one domain.

**Comment by orangejenny** [comment]
You might just not have gotten here yet, but looking at the ticket, this PR is also going to need to update the `create_usercases` task to create usercases for web users when a project starts using usercase.

We'll also need to create usercases for web users in projects that are using usercase now. Once this is released, projects are going to start seeing web user usercases appear ad hoc, as web users are saved. We might want to hold off on releasing this until usercases are available in the restore. Let's chat on the ticket, so the product team can participate.

File: corehq/apps/callcenter/tasks.py (hunk 11 above)
Comment (orangejenny) [reply corehq/apps/callcenter/tasks.py:49]: Looking at the test failures, checking `UNIT_TESTING` here will likely fix the one that says `Database queries to 'default' are not allowed in SimpleTestCase subclasses.` - that happens when you try to access the db in a test that doesn't allow it, which sounds like some test that didn't previously use the db is now trying to save usercases.

**Comment by orangejenny** [comment]
I keep thinking of additional things. [This](https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/tests/test_usercase_dbaccessor.py) would be a good test to update, to verify that usercases are created for web users.

File: corehq/apps/callcenter/sync_usercase.py [hunk 12]
```diff
@@ -226,32 +227,35 @@ def _call_center_location_owner(user, ancestor_level):
...
     This is also called to create usercases when the usercase is used for the
     first time.
     """
+    domain_obj = Domain.get_by_name(domain)
     with CriticalSection(get_sync_lock_key(user._id)):
         helpers = list(chain(
-            _iter_sync_usercase_helpers(user),
-            _iter_call_center_case_helpers(user),
+            _iter_sync_usercase_helpers(user, domain_obj),
+            _iter_call_center_case_helpers(user, domain_obj),
```
Comment (snopoke) [inline corehq/apps/callcenter/sync_usercase.py:258]:
```suggestion
            _iter_call_center_case_helpers(user, domain_obj) if isinstance(user, CommCareUser) else [],
```

(could also push this down to `_iter_call_center_case_helpers` which might be cleaner)

File: corehq/apps/callcenter/tasks.py [hunk 13]
```diff
@@ -46,6 +47,18 @@ def calculate_indicators():
...
+    if settings.UNIT_TESTING:
+        return
+    for domain in user.get_domains():
+        domain_obj = Domain.get_by_name(domain)
+        if domain_obj.usercase_enabled and domain in USH_WEB_USER_CASE_CREATION.get_enabled_domains():
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:55]: `get_enabled_domains` isn't cached to it should be moved outside of the `for` loop. Alternately you could use `USH_WEB_USER_CASE_CREATION.enabled(domain)`

File: corehq/apps/callcenter/tasks.py [hunk 14]
```diff
@@ -46,6 +47,18 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_web_user_usercases_if_applicable(user, spawn_task):
+    if settings.UNIT_TESTING:
+        return
+    for domain in user.get_domains():
+        domain_obj = Domain.get_by_name(domain)
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:54]: I think the settings check should move into the loop:

```suggestion
        domain_obj = Domain.get_by_name(domain)
        if not domain_obj and settings.UNIT_TESTING:
            continue
```

File: corehq/apps/callcenter/tasks.py [hunk 15]
```diff
@@ -58,12 +71,18 @@ def sync_usercases_if_applicable(user, spawn_task):
...
+    sync_usercases(user, domain)
 
 
-@task(serializer='pickle', queue='background_queue')
-def sync_usercases_task(user_id):
+@task(queue='background_queue')
+def sync_usercases_task(user_id, domain):
     user = CommCareUser.get_by_user_id(user_id)
-    sync_usercases(user)
+    sync_usercases(user, domain)
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:88]: These could be combined into a single tasks since `CouchUser.get_by_user_id` works for both web and mobile users
```python
@task(queue='background_queue')
def sync_usercases_task(user_id, domain):
    user = CouchUser.get_by_user_id(user_id)
    sync_usercases(user, domain)
```

File: corehq/form_processor/tests/test_usercase_dbaccessor.py [hunk 16]
```diff
@@ -28,8 +30,15 @@ def setUp(self):
...
                             update={'hq_user_id': self.user._id})
+        print(factory.create_case(case_type=USERCASE_TYPE, owner_id=self.web_user._id, case_name='web',
+                            update={'hq_user_id': self.web_user._id}))
 
     def test_get_usercase(self):
         usercase = self.accessor.get_case_by_domain_hq_user_id(self.user._id, USERCASE_TYPE)
         self.assertIsNotNone(usercase)
         self.assertEqual(usercase.name, 'bar')
+
+    def test_get_web_user_usercase(self):
```
Comment (snopoke) [inline corehq/form_processor/tests/test_usercase_dbaccessor.py:41]: I don't think this test is necessary since the db accessor isn't aware of the user type at all and only uses the user ID and case type for the lookup.

**Review by snopoke — COMMENTED** [review]
Still needs some tests for web users.

File: corehq/apps/callcenter/sync_usercase.py (hunk 12 above)
Comment (dominicochoa-dimagi) [reply corehq/apps/callcenter/sync_usercase.py:258]: This is because we don't want to support call center features for Web Users correct?

Comment (snopoke) [reply corehq/apps/callcenter/sync_usercase.py:258]: yes, that's correct.

**Comment by orangejenny** [comment]
Recent commits look good to me. Doesn't this still need to update `create_usercases` to create cases for web users?

**Comment by dominicochoa-dimagi** [comment]
Yes it does let me push that really quick.

File: corehq/apps/callcenter/sync_usercase.py [hunk 17]
```diff
@@ -179,27 +180,30 @@ def get_sync_lock_key(user_id):
...
-        yield _get_sync_usercase_helper(user, config.case_type, owner_id, case)
+        case, owner_id = _get_call_center_case_and_owner(user, domain_obj)
+        yield _get_sync_usercase_helper(user, domain_obj.name, config.case_type, owner_id, case)
 
 
 CallCenterCaseAndOwner = namedtuple('CallCenterCaseAndOwner', 'case owner_id')
 
 
-def _get_call_center_case_and_owner(user):
+def _get_call_center_case_and_owner(user, domain_obj):
```
Comment (orangejenny) [inline corehq/apps/callcenter/sync_usercase.py:201]: Now that `_iter_call_center_case_helpers` returns early for web users, can't you remove `domain_obj` from this function and go back to using `user.domain`?

File: corehq/apps/callcenter/tasks.py [hunk 18]
```diff
@@ -46,6 +47,17 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_web_user_usercases_if_applicable(user, domain, spawn_task):
+    domain_obj = Domain.get_by_name(domain)
+    if not domain_obj and settings.UNIT_TESTING:
```
Comment (orangejenny) [inline corehq/apps/callcenter/tasks.py:52]: Why bother checking for `domain_obj` here?

File: corehq/apps/callcenter/sync_usercase.py (hunk 12 above)
Comment (orangejenny) [reply corehq/apps/callcenter/sync_usercase.py:263]: Doesn't this mean `sync_call_center_user_case` could go back to only accepting mobile users and wouldn't need the domain passed in?

File: corehq/apps/users/models.py [hunk 19]
```diff
@@ -2433,6 +2433,12 @@ def get_dimagi_emails_by_domain(cls, domain):
             if user_doc['email'].endswith('@dimagi.com'):
                 yield user_doc['email']
 
+    def save(self, fire_signals=True, **params):
+        super().save(fire_signals=fire_signals, **params)
+        from corehq.apps.callcenter.tasks import sync_web_user_usercases_if_applicable
+        for domain in self.get_domains():
+            sync_web_user_usercases_if_applicable(self, domain, spawn_task=True)
```
Comment (orangejenny) [inline corehq/apps/users/models.py:2440]: `CommCareUser.save` only calls `sync_usercases_if_applicable` if `fire_signals` is true. I'd expect this to behave the same.

File: corehq/toggles.py [hunk 20]
```diff
@@ -849,6 +849,16 @@ def _ensure_valid_randomness(randomness):
     """
 )
 
+USH_WEB_USER_CASE_CREATION = StaticToggle(
+    'web_user_case_creation',
+    "USH: Enable the creation of user cases for web users.",
+    TAG_CUSTOM,
+    help_link='.',
```
Comment (orangejenny) [inline corehq/toggles.py:856]: This PR looks almost ready, so it'd be a good time to create docs and add the link here. You can make a child page of [this page](https://confluence.dimagi.com/pages/viewpage.action?pageId=41485226). It doesn't need to be long - [this page](https://confluence.dimagi.com/display/saas/USH%3A+Enable+Web+User+Activity+Report) is a basic example of docs for another feature flag that's USH-owned and still under development, like your new flag.

File: corehq/apps/callcenter/tasks.py (hunk 18 above)
Comment (dominicochoa-dimagi) [reply corehq/apps/callcenter/tasks.py:52]: I was being overly safe. For some reason if `Domain.get_by_name` returned None for some inexplicable reason this would catch it. In hindsight this is probably not necessary.

File: corehq/apps/app_manager/tasks.py [hunk 21]
```diff
@@ -18,8 +18,8 @@
 @task(queue='background_queue', ignore_result=True)
 def create_usercases(domain_name):
     from corehq.apps.callcenter.sync_usercase import sync_usercase
-    for user in CommCareUser.by_domain(domain_name):
-        sync_usercase(user)
+    for user in CouchUser.by_domain(domain_name):
```
Comment (snopoke) [inline corehq/apps/app_manager/tasks.py:21]: this will include web users even if the FF is not turned on:
```suggestion
    if USH_WEB_USER_CASE_CREATION.enabled(domain_name):
        users = CouchUser.by_domain(domain_name)
    else:
        users = CommCareUser.by_domain(domain_name)
    for user in users:
```

(code changes require importing the toggle)

File: corehq/apps/callcenter/tasks.py [hunk 22]
```diff
@@ -46,6 +47,17 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_web_user_usercases_if_applicable(user, domain, spawn_task):
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:50]: `spawn_task` is always True so can be removed

File: corehq/apps/callcenter/tasks.py [hunk 23]
```diff
@@ -46,6 +47,17 @@ def calculate_indicators():
...
+    domain_obj = Domain.get_by_name(domain)
+    if settings.UNIT_TESTING:
+        return
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:53]: Somehow this just feels wrong to me. This makes this code and any calling code un-testable. I know there's not much code here to test and that the functions being called are tested elsewhere. Since this is already limited by a FF I think it would be sufficient to just check the FF:

```python
    if not USH_WEB_USER_CASE_CREATION.enabled(domain):
        return
    domain_obj = Domain.get_by_name(domain)
    if domain_obj.usercase_enabled:
        if spawn_task:
            sync_usercases_task.delay(user._id, domain)
        else:
            sync_usercases(user, domain)
```

File: corehq/apps/callcenter/tasks.py (hunk 18 above)
Comment (orangejenny) [reply corehq/apps/callcenter/tasks.py]: Nah, you're right, it's better to be cautious.

**Comment by orangejenny** [comment]
Looks good to me! Leaving approval for @snopoke once his comments are addressed.

File: corehq/apps/callcenter/tasks.py (hunk 23 above)
Comment (dominicochoa-dimagi) [reply corehq/apps/callcenter/tasks.py:53]: should I also change this in `sync_usercases_if_applicable`? I was just going by what is there.
```python
def sync_usercases_if_applicable(user, spawn_task):
    if settings.UNIT_TESTING and not user.project:
        return
    if (user.project.call_center_config.enabled or user.project.usercase_enabled):
        if spawn_task:
            sync_usercases_task.delay(user._id, user.domain)
        else:
            sync_usercases(user, user.domain)
```

Comment (snopoke) [reply corehq/apps/callcenter/tasks.py:53]: no I don't think there's a need to change that since there is no FF enabling that feature

Comment (dominicochoa-dimagi) [reply corehq/apps/callcenter/tasks.py:53]: One thing to note, the `settings.UNIT_TESTING` check was added to prevent testing errors as a result of some `SimpleTestCases` that create a web user and then throw an error when the db accessor is called.

Comment (snopoke) [reply corehq/apps/callcenter/tasks.py:53]: as long as the FF is not enabled for those tests the it shouldn't be an issue

Comment (dominicochoa-dimagi) [reply corehq/apps/callcenter/tasks.py:53]: I'm now realizing this was an issue before I even created the FF so I'm mistaken.

So would something like this work:
```python
def sync_web_user_usercases_if_applicable(user, domain):
    domain_obj = Domain.get_by_name(domain)
    if not domain_obj:
        return
    if domain_obj.usercase_enabled and USH_WEB_USER_CASE_CREATION.enabled(domain):
        sync_usercases_task.delay(user._id, domain)
```

Comment (snopoke) [reply corehq/apps/callcenter/tasks.py:53]: I think the code should fail if the domain is `None`. The other code combines this with a check on unit testing setting so that it will fail if the condition occurs in production. My suggestion for this function is slightly different since this has a FF so we shouldn't need the settings check. Any unit test that is executing this code with the FF enabled should also be creating the domain object in the DB.

Let me know if you want to go to voice on this.

Comment (dominicochoa-dimagi) [reply corehq/apps/callcenter/tasks.py:53]: Okay that makes sense. Thanks for the clarification.
I think a quick 5 mins on voice would help tie a bow on this.

File: corehq/apps/callcenter/tasks.py [hunk 24]
```diff
@@ -47,15 +47,10 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
-def sync_web_user_usercases_if_applicable(user, domain, spawn_task):
+def sync_web_user_usercases_if_applicable(user, domain):
     domain_obj = Domain.get_by_name(domain)
-    if settings.UNIT_TESTING:
-        return
     if domain_obj.usercase_enabled and USH_WEB_USER_CASE_CREATION.enabled(domain):
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py]: the FF check should come before the `domain_obj` is fetched / used. The reason being that we can avoid an unnecessary object fetch and also in tests the domain may not exist so it the variable could be `None`. For tests with the FF enabled they would always create the domain object.

File: corehq/apps/callcenter/tasks.py [hunk 25]
```diff
@@ -46,6 +47,12 @@ def calculate_indicators():
         indicator_set.get_data()
 
 
+def sync_web_user_usercases_if_applicable(user, domain):
+    domain_obj = Domain.get_by_name(domain)
+    if USH_WEB_USER_CASE_CREATION.enabled(domain):
+        sync_usercases_task.delay(user._id, domain)
```
Comment (snopoke) [inline corehq/apps/callcenter/tasks.py:53]: this should still check the domain config if the FF is enabled:
```suggestion
    if not USH_WEB_USER_CASE_CREATION.enabled(domain):
        return
    domain_obj = Domain.get_by_name(domain)
    if domain_obj.usercase_enabled:
        sync_usercases_task.delay(user._id, domain)
```

File: corehq/form_processor/tests/test_usercase_dbaccessor.py [hunk 26]
```diff
@@ -4,7 +4,7 @@
 
 from corehq.apps.app_manager.const import USERCASE_TYPE
 from corehq.apps.domain.models import Domain
-from corehq.apps.users.models import CommCareUser
+from corehq.apps.users.models import CommCareUser, WebUser
```
Comment (snopoke) [inline corehq/form_processor/tests/test_usercase_dbaccessor.py:7]:
```suggestion
from corehq.apps.users.models import CommCareUser
```

**Review by snopoke — APPROVED** [review]

**Review by snopoke — APPROVED** [review]

File: corehq/apps/callcenter/sync_usercase.py [hunk 27]
```diff
@@ -226,31 +229,36 @@ def _call_center_location_owner(user, ancestor_level):
     return owner_id
 
 
-def sync_usercase(user):
+def sync_usercase(user, domain):
     with CriticalSection(get_sync_lock_key(user._id)):
```
Comment (esoergel) [inline corehq/apps/callcenter/sync_usercase.py:233]: Should this lock be user ID and domain?  I think as long as this code doesn't modify the user itself, that should be fine.  A task will be queued for each domain the user is a member of, and they'll all be fighting for the same lock.  I imagine we'd want to either let them all run in parallel (by adding domain to the lock key), or run them serially in a single task.

**Comment by orangejenny** [comment]
Confirming that the last few commits, related to renaming the feature flag, look fine.

**Comment by snopoke** [comment]
@dominicochoa-dimagi what's outstanding on this PR?

**Comment by dominicochoa-dimagi** [comment]
@snopoke this is ready. Just wanted okay for the new feature flag name. This has been tested on staging so it’s ready to merge otherwise.

File: corehq/apps/callcenter/sync_usercase.py (hunk 27 above)
Comment (snopoke) [reply corehq/apps/callcenter/sync_usercase.py:233]: @dominicochoa-dimagi I think this still needs to be addressed

Comment (dominicochoa-dimagi) [reply corehq/apps/callcenter/sync_usercase.py:233]: would this work: 
```python
def get_sync_lock_key(user_id, domain):
    return [f"sync_user_case_for_{user_id}_{domain}"]
```

then:

```python
def sync_usercase(user, domain):
    with CriticalSection(get_sync_lock_key(user._id, domain)):
        domain_obj = Domain.get_by_name(domain)
        _UserCaseHelper.commit(list(_iter_sync_usercase_helpers(user, domain_obj)))
```

File: corehq/apps/callcenter/sync_usercase.py [hunk 28]
```diff
@@ -176,12 +176,12 @@ def _not_same(val1, val2):
...
 
-def get_sync_lock_key(user_id):
-    return ["sync_user_case_for_%s" % user_id]
+def get_sync_lock_key(user_id, domain):
+    return [f"sync_user_case_for_{user_id}_{domain}"]
 
 
-def sync_call_center_user_case(user):
-    with CriticalSection(get_sync_lock_key(user._id)):
+def sync_call_center_user_case(user, domain):
```
Comment (orangejenny) [inline corehq/apps/callcenter/sync_usercase.py:183]: Not blocking: It looks like this is only used in tests. If so, it can be deleted.

**Review by orangejenny — APPROVED** [review]
Last commits look good. This is good to merge - I left one comment about possible dead code, but I'd suggest doing it in a followup PR.
